src/runtime/device_interface.cpp

```diff
@@ -509,54 +467,131 @@ WEAK int halide_buffer_copy(void *user_context, struct halide_buffer_t *src,
+    // no host buffer, will return an error. Some of these could be
+    // handled by allocating temporary host memory.
+    //
+    // It is assumed that if two device runtimes have copy compatible buffers
+    // both will handle a copy between their types of buffers.
+
+    // Give more descriptive names to conditions.
+    const bool from_device_valid = (src->device != 0) &&
+                                   (src->host == NULL || !src->host_dirty());
+    const bool to_device = dst_device_interface != NULL;
+    const bool to_host = dst_device_interface == NULL;
+    const bool from_host_exists = src->host != NULL;
+    const bool to_host_exists = dst->host != NULL;
+
+    // If a device to device copy is requested, try to do it directly.
+    err = halide_error_code_incompatible_device_interface;
+    if (from_device_valid && to_device) {
+        debug(user_context) << "halide_buffer_copy_already_locked: device to device case.\n";
+        err = dst_device_interface->impl->buffer_copy(user_context, src, dst_device_interface, dst);
     }
 
-    if (dst_device_interface) {
-        // Make the dst interface handle arbitrary src device
-        // interfaces (e.g. CUDA might know how to copy out of an
-        // OpenGL texture). If the dst device interface doesn't
-        // recognize the src device interface, it should ask the src
-        // device interface to copy to host memory in the *dst* buffer
-        // first (so that only the subset required is copied), and
-        // then do a copy_to_device from there.
-        err = dst_device_interface->impl->buffer_copy(user_context, src, dst_device_interface, dst);
-    } else if (src_device_interface) {
-        // The src device interface can handle copies to host
-        err = src_device_interface->impl->buffer_copy(user_context, src, dst_device_interface, dst);
-    } else {
-        // This is a host->host copy. The default implementation can
-        // handle this.
-        err = halide_default_buffer_copy(user_context, src, dst_device_interface, dst);
+    if (err == halide_error_code_incompatible_device_interface) {
+        // Return an error for a case that cannot make progress without a temporary allocation.
+        // TODO: go ahead and do the temp allocation.
+        if (!from_host_exists && !to_host_exists) {
+            debug(user_context) << "halide_buffer_copy_already_locked: failing due to need for temp buffer.\n";
+            return halide_error_code_incompatible_device_interface;
+        }
+
+        if (to_host) {
+            debug(user_context) << "halide_buffer_copy_already_locked: to host case.\n";
+            err = src->device_interface->impl->buffer_copy(user_context, src, NULL, dst);
+            // Return on success or an error indicating something other
+            // than not handling this case went wrong.
+            if (err == halide_error_code_incompatible_device_interface) {
```

Comment:
Should these be if (err != 0) ? What if the implementation throws some other error.

Reply:
Wait, nvm, this is handling a specific error.